jina/executors/indexers/keyvalue.py

```diff
@@ -1,161 +1,72 @@
-        return len(r) > 0
-
-
-class JsonPbIndexer(BasePbIndexer):
-    """Storing and querying protobuf chunk/document in JSON using gzip and Python dict. """
-
-    def get_query_handler(self):
-        r = {}
-        with gzip.open(self.index_abspath, 'rt') as fp:
-            for l in fp:
-                if l:
-                    tmp = json.loads(l)
-                    for k, v in tmp.items():
-                        r[k] = Parse(v, jina_pb2.Document())
-                        self._size += 1
-        return r
-
-    def _add(self, keys: Iterator['jina_pb2.Document'], *args, **kwargs):
-        """Add a JSON-friendly object to the indexer
-
-        :param obj: an object can be jsonified
-        """
-        keys = {k.id: k for k in keys}
-        json.dump(keys, self.write_handler)
-        self._size += len(keys)
-        self.write_handler.write('\n')
-
-
-class BinaryPbIndexer(BasePbIndexer):
-    """Storing and querying protobuf chunk/document in binary using gzip and Python dict.
-
-    .. note::
-        it often yields better performance than :class:`JsonPbIndexer`
-    """
-
-    mode = 'b'
+        return self.FileHandler(self.index_abspath, 'wb')
 
     def get_query_handler(self):
-        with gzip.open(self.index_abspath, 'rb') as fp:
-            tmp = fp.read()
-        if tmp.endswith(__binary_delimiter__):
-            tmp = tmp[:-len(__binary_delimiter__)]
-
-        r = {}
-        for l in tmp.split(__binary_delimiter__):
-            b = jina_pb2.Document()
-            b.ParseFromString(l)
-            r[b.id] = b
+        with open(self.index_abspath + '.head', 'rb') as fp:
+            tmp = np.frombuffer(fp.read(), dtype=np.uint64).reshape([-1, 4])
+            d = {r[0]: r[1:] for r in tmp}
+        return d
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._total_byte_len = 0
+        self._page_size = mmap.ALLOCATIONGRANULARITY
+
+    def add(self, docs: Iterator['jina_pb2.Document'], *args, **kwargs):
+        _start = 0
+        for d in docs:
+            s = d.SerializeToString()
+            l = len(s)  #: the length
+            p = int(_start / self._page_size) * self._page_size  #: offset of the page
+            r = _start % self._page_size  #: the reminder, i.e. the start position given the offset
+            self.write_handler.header.write(np.array((d.id, p, r, r + l), dtype=np.uint64).tobytes())
+            _start += l
+            self.write_handler.body.write(s)
             self._size += 1
-        return r
 
-    def _add(self, keys: Iterator['jina_pb2.Document'], *args, **kwargs):
-        """Add a object to the indexer
+    def query(self, key: int) -> Optional['jina_pb2.Document']:
 
-        :param obj: an object
-        """
-        for k in keys:
-            self.write_handler.write(k.SerializeToString() + __binary_delimiter__)
-            self._size += 1
+        pos_info = self.query_handler.get(key, None)
```

Comment:
Personally, I don't like the magic, how this calls `get_query_handler`. I have to look at how and why this is implemented as it is. From my point of view it should either be a `@property` (if the invocation is cheap) or a direct call to the `get_query_handler` if invocation is expensive.